src/circuit.rs

```diff
@@ -0,0 +1,594 @@
+        {
+            let next_layer_num_wires = if layer_index == output_layer_index {
+                self.num_outputs
+            } else {
+                self.layers[
+                    // index from the end because we are iterating self.layers in reverse
+                    output_layer_index - layer_index
+                    // next layer of the circuit is -1
+                    - 1
+                ]
+                .num_wires
+            };
+            // A single gate may receive contributions from multiple quads, so preallocate a vector
+            // of length matching the next layer's number of input wires, and accumulate quad
+            // outputs into that.
+            // Depending on the circuit compiler, it's possible that there will be unused elements
+            // in this vector.
+            let mut gate_outputs = vec![FE::ZERO; next_layer_num_wires.into()];
+
+            // Note which gates receive contributions from Z quads.
+            let mut z_gate_indexes = HashSet::new();
+
+            for (quad_index, quad) in layer.quads.iter().enumerate() {
+                // Evaluate this quad: look up its value in the constants table, then multiply that
+                // by the value of the input wires.
+                let quad_value: FE = self.constant(quad.const_table_index).context(format!(
+                    "constant missing in quad {quad_index} on layer {layer_index}"
+                ))?;
+
+                let left_wire = wires[layer_index]
+                    .get(usize::from(quad.left_wire_index))
+                    .ok_or_else(|| {
+                        anyhow!(
+                            "quad {quad_index} on layer {layer_index} contains left wire index {} \
+                            not present in previous layer of circuit {:?}",
+                            quad.left_wire_index,
+                            wires[layer_index],
+                        )
+                    })?;
+                let right_wire = wires[layer_index]
+                    .get(usize::from(quad.right_wire_index))
+                    .ok_or_else(|| {
+                        anyhow!(
+                            "quad {quad_index} on layer {layer_index} contains right wire index {} \
+                            not present in previous layer of circuit {:?}",
+                            quad.right_wire_index,
+                            wires[layer_index],
+                        )
+                    })?;
+
+                // Specification interpretation verification: if this quad is part of Z, then its
+                // input wires should be coming from a Z gate, and should always be zero.
+                if quad_value.is_zero().into() {
+                    z_gate_indexes.insert(usize::from(quad.gate_index));
+
+                    if !bool::from((*left_wire * right_wire).is_zero()) {
+                        panic!("product of input wires to a Z quad should be zero");
```

Comment:
This should be an error instead of a panic, since it can be triggered by an input that doesn't satisfy a circuit.